designs/2022-eslint-to-esm/README.md

```diff
@@ -0,0 +1,212 @@
+
+# convert eslint codebase to esm
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+Convert the ESLint codebase to ESM and publish the ESM version.
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+The Node.js community has been written commonjs for a long time, the same goes for ESLint. But with ESM introduced in ES6 and officially supported by Node.js v12, more and more projects are moving to ESM. Not long ago, we had converted a few repos under the ESLint organization to ESM:
+* @eslint/eslintrc
+* espree
+* eslint-scope
+* eslint-visitor-keys
+* @eslint/create-config
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+There are two main ways to achieve this:
+1. esm-only: write esm, publish esm
+```js
+{
+    "type": "module",
+    "exports": {
+        "./package.json": "./package.json",
+        ".": "./lib/api.js",
+        "./use-at-your-own-risk": "./lib/unsupported-api.js"
+    }
+}
+```
+2. dual-mode: compile esm to cjs, publish esm+cjs
+```js
+    "type": "module",
+    "exports": {
+        "./package.json": "./package.json",
+        ".": {
+            "import": "./lib/api.js",
+            "require": "./dist/api.cjs"
+        },
+        "./use-at-your-own-risk": {
+            "import": "./lib/unsupported-api.js",
+            "require": "./dist/unsupported-api.cjs",
+        }
+    }
+```
+
+Node.js ESM has a limitation: ESM cannot be required. Approach 2 provides a good compatibility for existing projects that rely on ESLint without requiring users to adopt ESM as well. In previous repos, dual-mode was mostly used (except for '@eslint/create-config').
+
+However, dual-mode also has a few disadvantages:
+1. dependencies. Some ESLint's dependencies have been (or will be) esm-only, which means we won't be able to upgrade them if we continue to support CJS. For new dependencies, we cannot use them if they are esm-only.
+2. Slower installation & larger volume usage, even though most users only use ESM, they still need to download CJS modules.
+3. Higher maintenance costs. You need to add a build process, test and publish ESM/CJS code.
+4. The CLIs do not support dual-mode.
+
+In the (very) long run, I would suggest that we eventually go with esm-only, but to give users more time to upgrade, as an intermediate state, we could use dual-mode in ESLint v9:
+
+* eslint v9:
+    - ESLint CLI: esm-only
+    - Node.js API: dual mode
+* eslint 1x (at the point when the majority of the ecosystem is ready to support it, likely to be v10/v11/...):
+    - ESLint CLI: esm-only
+    - Node.js API: esm-only
+
+## Toolings esm supports
+
+* mocha ✅
+* proxyquire ❌
```

Comment:
until the loaders API is no longer experimental, mocking of ESM deps isn't available.

Reply:
we can use the experimental loader, as it's only used in the tests.

Reply:
ah, maybe i misunderstood. *consumers* who rely on the ability to mock eslint are who i'm concerned about here.

Reply:
the list was the toolings that eslint was relying on. will  update to clarify, and add the mocking eslint issue in FAQ.